### PR TITLE
Fit generator

### PR DESCRIPTION
**Pull request by fchollet**
Add multi-threaded `fit_generator` method in models.

- allows to do real-time data augmentation in parallel to the model training
- allows to do out-of-core processing of large datasets without loss of efficiency
- allows to download batches of data for training on the fly without loss of efficiency (i.e. training on remote data)

etc.

### CODE REVIEW

**Comment by jeanpijon** [comment]
Hi, I have some contributions to this method.
* separate parallelization code from fit (ideal seems keras.utils.io_utils.py)
* validation should also be generator (if it is already in the memory, you can easily convert array -> generator)
* keras.model.{Sequential,Graph}.fit_generator shares lots of code, I would factor out all the closures.

What do you think - I have this code written, and it would take only a while to modify it for keras, but I wnat to know, if you would want to merge it afterwards.

Thanks
J

**Comment by fchollet** [comment]
You can submit a PR, I will look at it (maybe not right away). If you
already have the code written, then it doesn't involve any more work from
you.

**Comment by jeanpijon** [comment]
OK, I am on it.
